peps/pep-0759.rst

```diff
@@ -0,0 +1,472 @@
+
+This PEP defines what it means to be "safe" in this context, along with a new
+package upload file format called a ``.rim`` file. It defines how ``.rim``
+files affect the metadata returned for a package's `Simple Repository API
+<https://packaging.python.org/en/latest/specifications/simple-repository-api/>`__
+in both HTML and JSON format, and how traditional wheels can easily be turned
+into ``.rim`` files.
+
+Rationale
+=========
+
+The Python Package Index, hosted at pypi.org imposes `default limits
+<https://pypi.org/help/>`__ on upload artifact file size (100 MiB) and total
+project size (10.0 GiB). Most projects can comfortably fit within these limits
+during the lifetime of the project, through years of uploads. A few projects
+have encountered these limits, and have been granted both `file size
+exceptions <https://github.com/pypi/support/issues/4378>`__ and `project size
+exceptions <https://github.com/pypi/support/issues/4379>`__, allowing them to
+continue uploading new releases without having to take more drastic measures,
+such as removing files which may potentially still be in use by consumers
+(e.g. through version pins).
+
+A related workaround is the `"wheel stub"
+<https://github.com/wheel-next/wheel-stub>`__ approach, which provides an
+indirect link between PyPI and an external third party package index, where
+such limitations can be avoided. Wheel stubs are `source distributions
+<https://packaging.python.org/en/latest/specifications/source-distribution-format/>`__,
+a.k.a. "sdists" which utilize a :pep:`517` build backend that, instead of turning
+source code into a binary wheel, performs some logic to calculate the URL for
+an existing, externally hosted wheel to download.  Technically speaking, this
+satisfies the build backend contract, but the link between PyPI, the sdist,
+and the externally hosted wheel isn't evident, easily discoverable, or
+auditable.
+
+Historical context
+==================
+
+In 2013, :pep:`438` proposed a "backward-compatible two-phase transition
+process" to modify several aspects of release file hosting on PyPI. As this
+PEP describes, PyPI originally supported only project and release
+*registration* without also allowing for artifact file hosting. As such, most
+projects hosted release file artifacts elsewhere. Artifact hosting was later
+added, but the mix of externally and PyPI-hosted files lead to a wide range of
+usability and potential security-related problems. PEP 438 was an attempt to
+provide several facilities to allow external hosting while promoting a
+PyPI-first hosting preference.
+
+PEP 438 was complex, with three different "hosting modes", ``rel`` metadata in
+the simple HTML index pages to signify hosting locations, and a two-phase
+transition plan affecting PyPI and installer tools. PEP 438 was ultimately
+retracted in 2015 by :pep:`470`. In one respect, PEP 438 did succeed in:
+
+   bringing about more people to utilize PyPI's repository features, an
```

Comment:
This quote comes from PEP 470 but that's not quite evident; maybe make it more explicit.

Reply:
Thanks.  I think I've reworded this to make it more clear.